From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Wolfgang Bumiller <w.bumiller@proxmox.com>
Date: Thu, 7 May 2026 10:06:20 +0200
Subject: [PATCH 1/2] PVE: Revert "lxc/conf: drop cgroup1 config options
 (lxc.cgroup.*)"

This reverts commit f29c3a3a6e4b93100ad773c7cfb4098f6ba886e7.

This is done so that left-over config values from included configs
don't completely prevent containers from starting.
---
 src/lxc/conf.c    |  7 +++++++
 src/lxc/conf.h    |  1 +
 src/lxc/confile.c | 35 ++++++++++++++++++++++++++++++++++-
 3 files changed, 42 insertions(+), 1 deletion(-)

diff --git a/src/lxc/conf.c b/src/lxc/conf.c
index 2a81c70fd..5749cb35a 100644
--- a/src/lxc/conf.c
+++ b/src/lxc/conf.c
@@ -3103,6 +3103,7 @@ struct lxc_conf *lxc_conf_init(void)
 	new->rootfs.fd_path_pin = -EBADF;
 	new->rootfs.dfd_idmapped = -EBADF;
 	new->logfd = -1;
+	INIT_LIST_HEAD(&new->cgroup);
 	INIT_LIST_HEAD(&new->cgroup2);
 	/* Block ("allowlist") all devices by default. */
 	new->bpf_devices.list_type = LXC_BPF_DEVICE_CGROUP_ALLOWLIST;
@@ -4012,6 +4013,11 @@ int lxc_clear_cgroups(struct lxc_conf *c, const char *key, int version)
 		namespaced_token	= "lxc.cgroup2.";
 		namespaced_token_len	= STRLITERALLEN("lxc.cgroup2.");
 		list = &c->cgroup2;
+	} else if (version == CGROUP_SUPER_MAGIC) {
+		global_token		= "lxc.cgroup";
+		namespaced_token	= "lxc.cgroup.";
+		namespaced_token_len	= STRLITERALLEN("lxc.cgroup.");
+		list = &c->cgroup;
 	} else {
 		return ret_errno(EINVAL);
 	}
@@ -4259,6 +4265,7 @@ void lxc_conf_free(struct lxc_conf *conf)
 	free(conf->lsm_se_keyring_context);
 	lxc_seccomp_free(&conf->seccomp);
 	lxc_clear_config_caps(conf);
+	lxc_clear_cgroups(conf, "lxc.cgroup", CGROUP_SUPER_MAGIC);
 	lxc_clear_cgroups(conf, "lxc.cgroup2", CGROUP2_SUPER_MAGIC);
 	lxc_clear_cgroups_devices(conf);
 	lxc_clear_hooks(conf, "lxc.hook");
diff --git a/src/lxc/conf.h b/src/lxc/conf.h
index ea4e19940..762d58901 100644
--- a/src/lxc/conf.h
+++ b/src/lxc/conf.h
@@ -417,6 +417,7 @@ struct lxc_conf {
 	struct utsname *utsname;
 
 	struct {
+		struct list_head cgroup;
 		struct list_head cgroup2;
 		struct bpf_devices bpf_devices;
 	};
diff --git a/src/lxc/confile.c b/src/lxc/confile.c
index 0bebe4687..b313114ae 100644
--- a/src/lxc/confile.c
+++ b/src/lxc/confile.c
@@ -65,6 +65,7 @@ lxc_config_define(apparmor_profile);
 lxc_config_define(apparmor_raw);
 lxc_config_define(cap_drop);
 lxc_config_define(cap_keep);
+lxc_config_define(cgroup_controller);
 lxc_config_define(cgroup2_controller);
 lxc_config_define(cgroup_dir);
 lxc_config_define(cgroup_monitor_dir);
@@ -205,6 +206,7 @@ static struct lxc_config_t config_jump_table[] = {
 	{ "lxc.cgroup.dir.container",       true,  set_config_cgroup_container_dir,       get_config_cgroup_container_dir,       clr_config_cgroup_container_dir,       },
 	{ "lxc.cgroup.dir",                 true,  set_config_cgroup_dir,                 get_config_cgroup_dir,                 clr_config_cgroup_dir,                 },
 	{ "lxc.cgroup.relative",            true,  set_config_cgroup_relative,            get_config_cgroup_relative,            clr_config_cgroup_relative,            },
+	{ "lxc.cgroup",                     false, set_config_cgroup_controller,          get_config_cgroup_controller,          clr_config_cgroup_controller,          },
 	{ "lxc.console.buffer.size",        true,  set_config_console_buffer_size,        get_config_console_buffer_size,        clr_config_console_buffer_size,        },
 	{ "lxc.console.logfile",            true,  set_config_console_logfile,            get_config_console_logfile,            clr_config_console_logfile,            },
 	{ "lxc.console.path",               true,  set_config_console_path,               get_config_console_path,               clr_config_console_path,               },
@@ -1932,6 +1934,9 @@ static int __set_config_cgroup_controller(const char *key, const char *value,
 	if (version == CGROUP2_SUPER_MAGIC) {
 		token = "lxc.cgroup2.";
 		token_len = 12;
+	} else if (version == CGROUP_SUPER_MAGIC) {
+		token = "lxc.cgroup.";
+		token_len = 11;
 	} else {
 		return ret_errno(EINVAL);
 	}
@@ -1957,12 +1962,22 @@ static int __set_config_cgroup_controller(const char *key, const char *value,
 
 	new_cgroup->version = version;
 
-	list_add_tail(&new_cgroup->head, &lxc_conf->cgroup2);
+	if (version == CGROUP2_SUPER_MAGIC)
+		list_add_tail(&new_cgroup->head, &lxc_conf->cgroup2);
+	else
+		list_add_tail(&new_cgroup->head, &lxc_conf->cgroup);
 	move_ptr(new_cgroup);
 
 	return 0;
 }
 
+static int set_config_cgroup_controller(const char *key, const char *value,
+					struct lxc_conf *lxc_conf, void *data)
+{
+	return __set_config_cgroup_controller(key, value, lxc_conf,
+					      CGROUP_SUPER_MAGIC);
+}
+
 static int set_config_cgroup2_controller(const char *key, const char *value,
 					 struct lxc_conf *lxc_conf, void *data)
 {
@@ -3888,6 +3903,11 @@ static int __get_config_cgroup_controller(const char *key, char *retv,
 		namespaced_token = "lxc.cgroup2.";
 		namespaced_token_len = STRLITERALLEN("lxc.cgroup2.");
 		list = &c->cgroup2;
+	} else if (version == CGROUP_SUPER_MAGIC) {
+		global_token = "lxc.cgroup";
+		namespaced_token = "lxc.cgroup.";
+		namespaced_token_len = STRLITERALLEN("lxc.cgroup.");
+		list = &c->cgroup;
 	} else {
 		return ret_errno(EINVAL);
 	}
@@ -3914,6 +3934,13 @@ static int __get_config_cgroup_controller(const char *key, char *retv,
 	return fulllen;
 }
 
+static int get_config_cgroup_controller(const char *key, char *retv, int inlen,
+					struct lxc_conf *c, void *data)
+{
+	return __get_config_cgroup_controller(key, retv, inlen, c,
+					      CGROUP_SUPER_MAGIC);
+}
+
 static int get_config_cgroup2_controller(const char *key, char *retv, int inlen,
 					 struct lxc_conf *c, void *data)
 {
@@ -4904,6 +4931,12 @@ static inline int clr_config_keyring_session(const char *key,
 	return 0;
 }
 
+static inline int clr_config_cgroup_controller(const char *key,
+					       struct lxc_conf *c, void *data)
+{
+	return lxc_clear_cgroups(c, key, CGROUP_SUPER_MAGIC);
+}
+
 static inline int clr_config_cgroup2_controller(const char *key,
 						struct lxc_conf *c, void *data)
 {
-- 
2.47.3

